beacon-chain/db/kv/state.go

```diff
@@ -99,30 +114,112 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
+	validatorKeys := make([][]byte, len(states))              // For every state, this stores a compressed list of validator keys.
+	realValidators := make([][]*ethpb.Validator, len(states)) // It's temporary structure to restore state in memory after persisting it.
 	for i, st := range states {
 		pbState, err := v1.ProtobufBeaconState(st.InnerStateUnsafe())
 		if err != nil {
 			return err
 		}
+		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
+			// store the real validators to restore the in memory state later.
+			realValidators[i] = pbState.Validators
+
+			// yank out the validators and store them in separate table to save space.
+			var hashes []byte
+			for _, val := range pbState.Validators {
+				valBytes, encodeErr := encode(ctx, val)
+				if encodeErr != nil {
+					return encodeErr
+				}
+
+				// create the unique hash for that validator entry.
+				hash := hashutil.Hash(valBytes)
+				hashes = append(hashes, hash[:]...)
+
+				// note down the hash and the encoded validator entry
+				hashStr := hexutils.BytesToHex(hash[:])
+				validatorsEntries[hashStr] = valBytes
+			}
+
+			// zero out the validators List from the state bucket so that it is not stored as part of it.
+			pbState.Validators = nil
+
+			validatorKeys[i] = snappy.Encode(nil, hashes)
+		}
 		multipleEncs[i], err = encode(ctx, pbState)
 		if err != nil {
 			return err
 		}
 	}
 
-	return s.db.Update(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(stateBucket)
-		for i, rt := range blockRoots {
-			indicesByBucket := createStateIndicesFromStateSlot(ctx, states[i].Slot())
-			if err := updateValueForIndices(ctx, indicesByBucket, rt[:], tx); err != nil {
-				return errors.Wrap(err, "could not update DB indices")
+	if err := s.db.Update(func(tx *bolt.Tx) error {
+		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
+			bucket := tx.Bucket(stateBucket)
+			valIdxBkt := tx.Bucket(blockRootValidatorHashesBucket)
+			for i, rt := range blockRoots {
+				indicesByBucket := createStateIndicesFromStateSlot(ctx, states[i].Slot())
+				if err := updateValueForIndices(ctx, indicesByBucket, rt[:], tx); err != nil {
+					return errors.Wrap(err, "could not update DB indices")
+				}
+				if err := bucket.Put(rt[:], multipleEncs[i]); err != nil {
+					return err
+				}
+				if err := valIdxBkt.Put(rt[:], validatorKeys[i]); err != nil {
+					return err
+				}
+			}
+
+			// store the validator entries separately to save space.
+			valBkt := tx.Bucket(stateValidatorsBucket)
+			for hashStr, validatorEntry := range validatorsEntries {
+				key := hexutils.HexToBytes(hashStr)
+				// if the entry is not in the cache and not in the DB,
+				// then insert it in the DB and add to the cache.
+				if _, ok := s.validatorEntryCache.Get(key); !ok {
+					validatorEntryCacheMiss.Inc()
+					if valEntry := valBkt.Get(key); valEntry == nil {
```

Comment:
+1 on the Has() or checking the internals of Put()

Reply:
Again bolt doesn't seem to have Has(). 
More context here https://github.com/boltdb/bolt/issues/321#issuecomment-76822462